File: src/jastudio_src/jastudio/note/anki_backend_note_creator.py
from __future__ import annotations

from typing import TYPE_CHECKING

from anki.notes import Note
from jastudio.ankiutils import app
from JAStudio.Core.Note import NoteTypes
from jastudio.note import studing_status_helper
from jastudio.note.jpnotedata_shim import JPNoteDataShim

if TYPE_CHECKING:
    from collections.abc import Callable

    from JAStudio.Core.Note import JPNote, KanjiNote, SentenceNote, VocabNote


class AnkiBackendNoteCreator:
    @classmethod
    def _create_note(cls, note: JPNote, note_type: str, callback: Callable[[], None]) -> None:
        backend_note = Note(app.anki_collection(), app.anki_collection().models.by_name(note_type))
        JPNoteDataShim.sync_note_to_anki_note(note, backend_note)
        app.anki_collection().addNote(backend_note)
        # Register the Anki ID → domain NoteId mapping.
        # The note already has its domain NoteId from construction.
        note.UpdateInCache()
        callback()
        studing_status_helper.update_note_in_studying_cache(backend_note)

    def create_kanji(self, note: KanjiNote, callback: Callable[[], None]) -> None:
        self._create_note(note, NoteTypes.Kanji, callback)

    def create_vocab(self, note: VocabNote, callback: Callable[[], None]) -> None:
        self._create_note(note, NoteTypes.Vocab, callback)

    def create_sentence(self, note: SentenceNote, callback: Callable[[], None]) -> None:
        self._create_note(note, NoteTypes.Sentence, callback)

File: src/src_dotnet/JAStudio.Anki.PythonInterop/AnkiBackendNoteCreator.cs
using System;
using JAStudio.Core.Note;
using JAStudio.Core.Note.Sentences;
using JAStudio.Core.Note.Vocabulary;
using JAStudio.PythonInterop.Utilities;

namespace JAStudio.Anki.PythonInterop;

public class AnkiBackendNoteCreator : IBackendNoteCreator
{
   readonly PythonObjectWrapper _noteCreator;

   public AnkiBackendNoteCreator()
   {
      using(PythonEnvironment.LockGil())
      {
         _noteCreator = PythonEnvironment.Import("jastudio.note.anki_backend_note_creator")
            .Use(module => new PythonObjectWrapper(module.AnkiBackendNoteCreator()));
      }
   }

   public void CreateKanji(KanjiNote note, Action callback) => _noteCreator.Use(it => it.create_kanji(note));

   public void CreateVocab(VocabNote note, Action callback) => _noteCreator.Use(it => it.create_vocab(note));

   public void CreateSentence(SentenceNote note, Action callback) => _noteCreator.Use(it => it.create_sentence(note));
}
